Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicHandler.cpp

```diff
@@ -719,26 +727,51 @@ namespace AzToolsFramework
-                    *parentEntityDomInOwningTemplate, parentEntityDomAfterAddingEntity, parentId, undoBatch.GetUndoBatch());
-            }
+                // Add entity DOM in owning template.
+                PrefabDom newEntityDom;
+                m_instanceToTemplateInterface->GenerateDomForEntity(newEntityDom, *entity);
 
-            // Select the new entity (and deselect others).
-            AzToolsFramework::EntityIdList selection = {entityId};
+                TemplateId targetTemplateId;
+                PrefabUndoAddEntity::NewEntityInfo newEntityInfo(entityId, newEntityDom);
+                PrefabUndoAddEntity::ParentEntityInfo parentEntityInfo(parentId, *parentEntityDomInOwningTemplate, parentEntityDomAfterAddingEntity);
+                AZStd::string entityAliasPathPrefix;
 
-            SelectionCommand* selectionCommand = aznew SelectionCommand(selection, "");
-            selectionCommand->SetParent(undoBatch.GetUndoBatch());
+                if (IsPrefabOverridesUxEnabled() && !m_prefabFocusPublicInterface->IsOwningPrefabBeingFocused(parentId))
+                {
+                    // Climb up the instance hierarchy from the owning instance until it hits the focused prefab instance.
+                    InstanceClimbUpResult climbUpResult = m_prefabFocusHandler.ClimbUpToFocusedOrRootInstanceFromEntity(parentId);
+                    if (!climbUpResult.m_isTargetInstanceReached)
+                    {
+                        return AZ::Failure(AZStd::string::format(
+                            "Parent entity (id: '%llu') is not owned by a descendant of the focused prefab instance.",
+                            static_cast<AZ::u64>(parentId)));
+                    }
 
-            ToolsApplicationRequests::Bus::Broadcast(&ToolsApplicationRequests::SetSelectedEntities, selection);
+                    LinkId linkId = climbUpResult.m_climbedInstances.back()->GetLinkId();
+                    LinkReference link = m_prefabSystemComponentInterface->FindLink(linkId);
+                    if (link == AZStd::nullopt)
+                    {
+                        return AZ::Failure(AZStd::string::format(
+                            "Can't find link (id: '%llu') in prefab system.", static_cast<AZ::u64>(linkId)));
+                    }
+                    targetTemplateId = link->get().GetTargetTemplateId();
+
+                    entityAliasPathPrefix = PrefabInstanceUtils::GetRelativePathFromClimbedInstances(climbUpResult.m_climbedInstances);
+                }
+                else
+                {
+                    targetTemplateId = entityOwningInstance.GetTemplateId();
+                }
 
-            // Add entity DOM in owning template.
-            PrefabDom newEntityDom;
-            m_instanceToTemplateInterface->GenerateDomForEntity(newEntityDom, *entity);
+                PrefabUndoHelpers::AddEntity(
+                    AZStd::move(parentEntityInfo),
+                    AZStd::move(newEntityInfo),
```

Comment:
I am not so sure how it works if we move PrefabDomValue reference. I feel like a better model would be:
1. ParentEntityInfo owns the two PrefabDoms. Upon creation, they copy from the original DOMs.
2. The public function owns ParentEntityInfo. And AddEntity() takes in ParentEntityInfo as a const reference. No move is needed here.
3. Apply to EntityInfo as well.

This looks more clear to me and easier to debug memory issue if there is.

Reply:
That solution would require copying the doms though. I'm not sure if its a good idea. We only need the doms in parentEntityInfo and newEntityInfo to be valid until the undo and redo patches are created. So they are short lived. So copying them into their own doms would become a throw-away operation.